### PR TITLE
Fix Address deserialization compatibility

### PR DESCRIPTION
**Pull request by lehins**
# Description

Starting with Babbage era it is not possible to decode a TxOut with a bad address, such as one that had trailing junk data or a Ptr address with values that are too large or trailed with junk. That being said, prior to Babbage there was a few Ptrs that violated those restrictions and thus need to be supported. This PR fixes this backwards compatibility of retaining buggy behavior for older eras:

* Implement lenient Ptr decoding
* Improve BabbageTxOut serialization

# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] Any changes are noted in the [changelog](https://github.com/input-output-hk/cardano-ledger/blob/master/CHANGELOG.md)
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (which can be run with `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (which can be done with `scripts/cabal-format.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs [hunk 1]
```diff
@@ -791,13 +808,24 @@ decodeVariableLengthWord32 name buf = do
         res <- decode7BitVarLength name buf (\_ -> failDecoding name "too many bytes.") acc
         -- Only while decoding the last 7bits we check if there was too many
         -- bits supplied at the beginning.
-        unless (bufUnsafeIndex buf off0 .&. 0b01110000 == 0) $
+        unless (bufUnsafeIndex buf off0 .&. 0b11110000 == 0b10000000) $
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs:811]: Change to this check is not needed, this is mostly to clarify that the most significant bit is always set. In other words it is impossible for it to not be set at this particular point.

File: libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs [hunk 2]
```diff
@@ -791,13 +808,24 @@ decodeVariableLengthWord32 name buf = do
...
           failDecoding name "More than 32bits was supplied"
         pure res
       {-# INLINE d7last #-}
   d7 (d7 (d7 (d7 d7last))) 0
 {-# INLINE decodeVariableLengthWord32 #-}
 
+-- | This decoder is here only with the purpose of preserving old buggy behavior. Should
+-- not be used for anything else.
```
Comment (JaredCorduan) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs:819]: want to give it a scarier name?

File: libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs [hunk 3]
```diff
@@ -291,12 +290,10 @@ isBootstrapRedeemer (BootstrapAddress (Byron.Address _ _ Byron.ATRedeem)) = True
...
-putPtr (Ptr slot (TxIx txIx) (CertIx certIx)) = do
-  putSlot slot
-  putVariableLengthWord64 ((fromIntegral @_ @Word64) txIx)
-  putVariableLengthWord64 ((fromIntegral @_ @Word64) certIx)
-  where
-    putSlot (SlotNo n) = putVariableLengthWord64 n
+putPtr (Ptr (SlotNo slot) (TxIx txIx) (CertIx certIx)) = do
+  putVariableLengthWord64 slot
+  putVariableLengthWord64 txIx
+  putVariableLengthWord64 certIx
```
Comment (JaredCorduan) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs:296]: lovely :)

**Review by JaredCorduan — APPROVED** [review]
LGTM! so nice that this is all the past now :relieved:

File: libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs (hunk 2 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Address.hs:819]: Nah, its alright. It is an internal function that is restricted to this module only. So the danger of anyone using it is almost non-existent.

**Pull request merged**